[WIP] CI Build wheels for arm64 on circleci.

cmarmo opened the pull request:
#### Reference Issues/PRs
Follow up of #20460

#### What does this implement/fix? Explain your changes.
This pull request generates wheels for python 3.7, 3.8, 3.9 on arm64 architectures when the commit message contains [cd build].

cmarmo committed:
[cd build] Build wheels for arm64 on circleci.

cmarmo commented:
Should I remove the travis build in this same pull request?
Also, I suppose the artifacts should be uploaded to Pypi at some time: I've checked #19060, but I will not have bandwidth for understanding how it works in the next weeks. It would be nice if the upload could be addressed in another pull request. Thank you very much for your understanding.

ogrisel commented:
> Should I remove the travis build in this same pull request?

I believe so. They currently error anyway.

The manually triggered github actions config that uploads the wheels stored in the anaconda.org staging area to PyPI should not be impacted by this PR.

However we will need to adapt the code that uploads the generated wheels to the anaconda.org staging area. In particular, I will have to configure the secrets environment variables with the anaconda.org tokens to have them available on our circle ci account.

ogrisel commented on .circleci/config.yml:
```diff
@@ -158,6 +158,11 @@ jobs:
+      - store_artifacts:
+          path: wheelhouse
+          destination: wheelhouse
+
```
I think it would make the CI status easier to understand to have a dedicated `linux-arm64-wheels` job (to be enabled only when `[cd build]` is active) in the circle ci config in addition to the `linux-arm64` entry should always run, independently of `[cd build]`.

ogrisel commented:
For information I have configured the `SCIKIT_LEARN_STAGING_UPLOAD_TOKEN` and `SCIKIT_LEARN_NIGHTLY_UPLOAD_TOKEN` environment variables on Circle CI.

They should automatically be available for this build config but only on the merged pull request (not in the builds that happen when pushing commits to a PR for security reasons).

cmarmo committed:
Remove travis.

Commit:
Merge branch 'main' into circle-cd-build-arm

cmarmo committed:
Move condition in circleci config.

cmarmo committed:
Fix parameter default.

cmarmo committed:
Fix string.

cmarmo committed:
Fix matches condition.

cmarmo committed:
Add wheels.

cmarmo committed:
Add script to build wheels

cmarmo committed:
Move conditions in workflows.

cmarmo committed:
[cd build] Trig wheels.

cmarmo committed:
Rename wheel script. Fix pattern.

cmarmo committed:
[cd build] Trig wheels.

cmarmo committed:
[cd build]Redefine regexp.

cmarmo committed:
[cd build]Redefine regexp.

cmarmo committed:
[cd build] Redefine regexp.

cmarmo committed:
[cd build] Fix version.

cmarmo committed:
Move when condition to job.

cmarmo committed:
Fix config.

cmarmo committed:
Move condition back to workflow.

cmarmo committed:
Set parameter.

cmarmo committed:
Remove unuseful definition.

cmarmo committed:
Revert.

cmarmo committed:
Remove condition.

cmarmo committed:
Fix.

cmarmo committed:
Fix again.

cmarmo committed:
Fix again.

cmarmo committed:
[cd build] Trig wheels.

cmarmo committed:
[cd build] Trig wheels.

cmarmo committed:
[cd build] Fix again.

cmarmo committed:
[cd build] Fix unexpected EOF.

cmarmo committed:
[cd build] Fix unexpected EOF.

cmarmo commented:
@ogrisel, I think I have addressed your comments. I have experimented a bit with [conditional workflows](https://support.circleci.com/hc/en-us/articles/360043638052-Conditional-steps-in-jobs-and-conditional-workflows) without success, that's why this took some time more than expected. Thanks for your patience.

alfaro96 commented on build_tools/github/check_wheels.py:
```diff
@@ -20,15 +20,6 @@
 # plus one more for the sdist
 n_wheels += 1
 
-# aarch64 builds from travis
-travis_config_path = Path.cwd() / ".travis.yml"
-with travis_config_path.open("r") as f:
-    travis_config = yaml.safe_load(f)
-
-jobs = travis_config["jobs"]["include"]
-travis_builds = [j for j in jobs if any("CIBW_BUILD" in env for env in j["env"])]
-n_wheels += len(travis_builds)
-
```
I think that we have to include here the number of wheels for circleci.

Or should this be done in a different PR?

cmarmo replied:
That was my first guess then I read [this comment](https://github.com/scikit-learn/scikit-learn/pull/19060#issuecomment-749600129) and I was lost... @ogrisel, please let us know, what should be done here. Thanks!

ogrisel replied:
We do need to count the 3 linux arm64 wheels in the `build_tools/github/check_wheels.py ` script prior to uploading everything to pypi.

ogrisel replied:
The point of [my comment](https://github.com/scikit-learn/scikit-learn/pull/19060#issuecomment-749600129) was to avoid counting the linux arm64 wheels twice in 2 almost concurrently merged PRs.

cmarmo replied:
Thanks, this wasn't clear to me.

cmarmo replied:
Thanks @alfaro96 for bringing this up. I have added the check. I'm resolving this conversation. Feel free to comment in the new version of the file.

ogrisel commented on build_tools/circle/build_wheels_arm.sh:
```diff
@@ -0,0 +1,18 @@
+sudo add-apt-repository --remove ppa:ubuntu-toolchain-r/test
+sudo apt-get update
+sudo apt-get install python3-virtualenv ccache
+python3 -m virtualenv --system-site-packages --python=python3 testenv
```
I don't think we need any system-level python package for this, no? We just need a bare-bone python to launch cibuildwheel itself.

```suggestion
python3 -m virtualenv --python=python3 testenv
```

cmarmo replied:
fixed.

ogrisel commented on build_tools/circle/build_wheels_arm.sh:
```diff
@@ -0,0 +1,18 @@
+    pywheel=`echo "cp"$pyversion"-manylinux_aarch64"`
+    export CIBW_BUILD=$pywheel
+    python -m pip install cibuildwheel
+        python -m cibuildwheel --output-dir wheelhouse
```
Extra indentation:

```suggestion
    python -m cibuildwheel --output-dir wheelhouse
```

ogrisel commented on build_tools/circle/build_wheels_arm.sh:
```diff
@@ -0,0 +1,18 @@
+python3 -m virtualenv --system-site-packages --python=python3 testenv
+source testenv/bin/activate
+pip install --upgrade pip
+for pyversion in 37 38 39; do
```
I am afraid that building the ARM wheels sequentially will slow down the release process. Actually, for 3 python, circle ci seems to be able to complete in 25m which is not too bad. Maybe this is because of ccache?

Still we might want to have one CI job per Python version and leverage Circle CI parallelism https://circleci.com/docs/2.0/parallelism-faster-jobs/. Although one would have to check how many concurrent workers are offered by default in the free Circle Ci account.

ogrisel replied:
Actually I am confused when reading the logs of the cibuildwheel: it seems that it's always using Python3.8 to build all the 3 wheel versions. This means that we will build 3.7 wheels against the Python 3.8 ABI.

I think we need to make sure that each Python version is run in its own circleci job config within an image with the matching Python version: https://circleci.com/docs/2.0/circleci-images/#python (3.7, 3.8 and 3.9).

ogrisel replied:
Unfortunately, according to https://circleci.com/pricing/ I don't think we can benefit from parallelism. So no need to try to get this to work.

However we still need to split the for loop into independent jobs to make sure we use the right Python version for each generated wheel.

cmarmo replied:
> I think we need to make sure that each Python version is run in its own circleci job config within an image with the matching Python version: https://circleci.com/docs/2.0/circleci-images/#python (3.7, 3.8 and 3.9).

If I understand correctly those customized python images are not available for arm (see the [available images for arm documentation](https://circleci.com/docs/2.0/arm-resources/#overview)), this means to reinstall a conda environment for each wheel on the host image `ubuntu-2004:202101-01`. 

> Actually I am confused when reading the logs of the cibuildwheel: it seems that it's always using Python3.8 to build all the 3 wheel versions. This means that we will build 3.7 wheels against the Python 3.8 ABI.

I am confused too then... :) ... I was thinking that the `CIBW_BUILD` variable was requesting a specific python version in the `manylinux` image... eg [this line](https://app.circleci.com/pipelines/github/scikit-learn/scikit-learn/16867/workflows/66cbe705-61df-4d38-b6f4-593ab0623498/jobs/148909?invite=true#step-102-551) shows that python 3.9 is called ... but it is not clear to me if it is during the build of the wheel... perhaps I should also add a bit of verbosity during the WIP state.

Anyway, I understand your concern, let me create the conda envs then.

ogrisel replied:
Yes you are right! For some reason I thought cibuildwheel used docker only for the repair step but it's indeed actually also used for the build step: https://github.com/pypa/cibuildwheel/blob/144733dc60fe0c9c3114e7ef2e4996ad919fec5c/cibuildwheel/linux.py#L186

Too bad that the `which python` before in the log does not print its output in the log, that would have lifted the ambiguity.

cmarmo replied:
I have configured `CIBW_BUILD_VERBOSITY: 1` as for the other wheels. Hope this will help.

ogrisel commented on build_tools/circle/build_wheels_arm.sh:
```diff
@@ -0,0 +1,18 @@
+set -e
+set -x
+
+sudo add-apt-repository --remove ppa:ubuntu-toolchain-r/test
```
Why is this PPA needed?

cmarmo replied:
euh.... this command remove a repository from the apt-repository... I used the azure ubuntu installation as model and it was used [there](https://github.com/scikit-learn/scikit-learn/blob/fac31e727947ad53f2ed107f58a10b56b165cee7/build_tools/azure/install.sh#L67) ... I have removed it.

ogrisel commented on build_tools/circle/build_wheels_arm.sh:
```diff
@@ -0,0 +1,18 @@
+
+sudo add-apt-repository --remove ppa:ubuntu-toolchain-r/test
+sudo apt-get update
+sudo apt-get install python3-virtualenv ccache
```
Once each build is running in its own circle ci docker image, we can just use the Python version pre-installed in the image instead and we would not need to create a venv anymore.

ogrisel replied:
I think we can also skip the ccache installation since the compiling happens inside isolated docker containers that won't use any host system command.

cmarmo replied:
done

cmarmo replied:
I kept the venv, as in ubuntu the default `python` command is `python3` , and I found the venv suite of commands more general... but no strong opinion on that.

ogrisel replied:
I agree.

ogrisel reviewed (commented):
Thank @cmarmo for the PR, is a first pass of comments. My main concern is that it seems that we are not building the 3.7 and 3.9 wheels with the matching Python version.

ogrisel commented on build_tools/circle/build_wheels_arm.sh:
```diff
@@ -0,0 +1,18 @@
+pip install --upgrade pip
+for pyversion in 37 38 39; do
+    pywheel=`echo "cp"$pyversion"-manylinux_aarch64"`
+    export CIBW_BUILD=$pywheel
```
We also need to configure cibuildwheel to run the scikit-learn tests on the generated wheel by defining the `CIBW_TEST_REQUIRES` and `CIBW_TEST_COMMAND` variables as done in:

https://github.com/scikit-learn/scikit-learn/blob/61ad2657a86dba20e15e4d76f58322d9c36ecc9e/.github/workflows/wheels.yml#L102-L103

cmarmo replied:
done

ogrisel commented:
Thanks for your reply in https://github.com/scikit-learn/scikit-learn/pull/20711#discussion_r700597285 @cmarmo. I think we can keeping a sequential loop in the same image for now. However we need to run the tests in a subsequent step to make sure that the generated wheels are valid.

ogrisel commented:
For your information, I downloaded the generated artifacts from https://app.circleci.com/pipelines/github/scikit-learn/scikit-learn/16867/workflows/66cbe705-61df-4d38-b6f4-593ab0623498/jobs/148909/artifacts and tried to run the test on a clean Python 3.7 env on a linux arm64 docker container on my local machine and all tests pass! \o/

cmarmo committed:
Merge branch 'main' into circle-cd-build-arm

cmarmo committed:
Add the number of new wheels. Add wheel tests. Address comments.

cmarmo committed:
[cd build] Trig wheel build.

ogrisel commented on build_tools/github/check_wheels.py:
```diff
@@ -17,18 +17,12 @@
 # 2 windows wheel (win32 + wind_amd64)
 n_wheels = 7 * n_python_versions
 
+# plus three more for arm64 on circleci
```
```suggestion
# plus three more for linux/arm64 on circleci
```

as I expect that we will be able to add macos/arm64 on github actions soon (with the help of docker/qemu).

ogrisel committed:
Update build_tools/github/check_wheels.py

ogrisel reviewed (commented):
I reviewed the log of the last `[cd build]` commit of this PR and it seems good:

https://app.circleci.com/pipelines/github/scikit-learn/scikit-learn/17282/workflows/072c9e95-8a75-4a9f-b1c2-fe018b6f26fc/jobs/150571

pytest runs successfully for the 3 python versions.

We just need to add the anaconda.org upload step as done on github actions:

https://github.com/scikit-learn/scikit-learn/blob/cfc1695dec0cac02a5bfce3e3bb53d6dafcc6a37/.github/workflows/wheels.yml#L165-L171

this is needed to upload the release artifacts to the staging area:

https://anaconda.org/scikit-learn-wheels-staging/scikit-learn/files

The staging area is then used by the final [publish_pypi](https://github.com/scikit-learn/scikit-learn/blob/main/.github/workflows/publish_pypi.yml) github action triggered manually in the github UI.

Edit: note that I have already added the necessarily secret upload tokens as environment variables on Circle CI. Those environment variables are only defined on jobs run from the commit pushed on branches of the main repo, not on commit pushed to PRs for security reasons. We just have to skip the upload to anaconda.org on PR-triggered builds (as is the case on github actions).

ogrisel commented:
@cmarmo to be able to reuse the `build_tools/github/upload_anaconda.sh` script, we need a working conda setup on this circle ci job. Let me give it a shot.

ogrisel committed:
MAINT setting up staging area upload step [cd build]

alfaro96 commented on build_tools/github/upload_anaconda.sh:
```diff
@@ -3,6 +3,14 @@
+conda install -y anaconda-client
+which anaconda
+
+# TODO: how to detect nightly builds for ARM on Circle CI?
```
Maybe we can use `"$GITHUB_EVENT_NAME" == "schedule" or "$CIRCLE_BRANCH" == "main"`?

ogrisel replied:
I was thinking on something like this. We might want to design a uniform strategy that would work for any CI. Let's keep the nightly build config for a follow-up PR.

alfaro96 replied:
+1.

alfaro96 commented on build_tools/github/upload_anaconda.sh:
```diff
@@ -11,12 +19,10 @@ else
-# Force a replacement if the remote file already exists
-anaconda -t $ANACONDA_TOKEN upload --force -u $ANACONDA_ORG dist/artifact/*
-echo "Index: https://pypi.anaconda.org/$ANACONDA_ORG/simple"
+if [ "$ANACONDA_TOKEN" == "" ];
```
I would prefer `if [ -z "$ANACONDA_TOKEN" ]`, but the proposed solution would work anyway.

alfaro96 reviewed (commented):
Thank you @ogrisel, just a few comments.

alfaro96 commented on .circleci/config.yml:
```diff
@@ -159,6 +159,28 @@ jobs:
+      - store_artifacts:
+          path: wheelhouse
+          destination: wheelhouse
+      - run: ./build_tools/github/upload_anaconda.sh
```
```suggestion
      - run: bash build_tools/github/upload_anaconda.sh
```

ogrisel committed:
Try to fix the upload_anaconda.sh call

Co-authored-by: Juan Carlos Alfaro Jiménez <JuanCarlos.Alfaro@uclm.es>

ogrisel committed:
[cd build]

ogrisel committed:
Define  in circle ci to be homegenous with github actions [cd build]

ogrisel committed:
bash syntax error [cd build]

ogrisel commented:
pfff, I hate this. It seems to work though, I might re-enable the full build.

However I have the feeling that using circle ci to sequentially build and test all the linux/arm64 wheels is going to be too slow on a single free circle ci executor. Maybe we should manage to find a way to fix the travis linux/arm64  wheel build instead.

ogrisel committed:
Reenable full linux/arm64 wheel [cd build]ing

ogrisel commented:
See concurrent experiments in #20958.

alfaro96 commented:
> pfff, I hate this. It seems to work though, I might re-enable the full build.
> 
> However I have the feeling that using circle ci to sequentially build and test all the linux/arm64 wheels is going to be too slow on a single free circle ci executor. Maybe we should manage to find a way to fix the travis linux/arm64 wheel build instead.

I have been a little bit out these months, what is happening with the `travis` wheel builder?

cmarmo commented:
Hi everybody, it seems that this PR is suddenly becoming urgent?
I was indeed working on how to select nightly builds OR [cd build]. But apparently [this is not necessary](https://github.com/scikit-learn/scikit-learn/pull/20711#discussion_r703014909)?
My opinion is that this PR could be merged without the uploading artifact step (then without the check in `check_builds`).
Small steps allow sometimes to move forward with less pain.
This implies that wheels for linux arm64 should be uploaded manually ... this is not the only thing manually done for the release....

Anyway, let me know if I can still do something or you prefer take over. Thanks.

ogrisel commented:
> I was indeed working on how to select nightly builds OR [cd build]. But apparently this is not necessary?

This is not needed for the 1.0 release we are preparing. But would be useful later.

> My opinion is that this PR could be merged without the uploading artifact step (then without the check in check_builds).

The upload step should work now and is needed to get the release automation working. Apparently there is just a missing PATH in my last commit:

```
WARNING: The script cibuildwheel is installed in '/home/circleci/mambaforge/bin' which is not on PATH.
```

I will push a fix.

My main concern however is that the free account circleci has only one executor that is already under heavy load to build the doc when when we merge PRs (especially at release time). So I was trying concurrently to fix the linux/arm64 wheel building config on travis to avoid loading circleci to avoid making the CI even slower to iterate with, especially during a release process.

ogrisel commented:
> I have been a little bit out these months, what is happening with the travis wheel builder?

It seems to be caused by output buffering issues by travis itself. I am trying to find a workaround in #20958.

ogrisel committed:
Try to expaned HOME variable in circleci config [cd build]

ogrisel committed:
This PATH and CONDA vars should be defined correctly [cd build]

ogrisel commented:
This Circle CI config probably needs tuning of the number of OPENBLAS/OMP_NUM_THREADS and build test pytest parallelism. However since we only have 1 executor, it will never be as fast as the travis build that I have just fixed in #20958: it only take less than 9min because of the job level parallelism and inner tuned parallelism.

So I would be in favor of merging #20958 and closing this (#20711).

We can later clean-up the travis config in follow-up PRs but at least the release process should be working and fast now.

ogrisel commented:
@glemaitre suggested that we could keep using circleci but only once a day for the schedule nightly job (to upload the wheels for the main branch to https://anaconda.org/scipy-wheels-nightly/scikit-learn/files every night. Build time is not a problem since it's not an interactive process.

And we could keep dedicate the travis config only for the release process with the manual `[cd build]` triggers to the main or release branches.

This way we could spare the limited free travis credits to dedicate them to release efforts.

If we agree on this plan, we would need to:

-  undelete the travis config in this PR;
- configure a cron job for this PR and disable the `[cd build]` manual trigger;
- it would still be worth optimizing the parallel environment variables to try to speed-up the tests (e.g. as done in #20958) in particular: `SKLEARN_BUILD_PARALLEL`, `OMP/OPENBLAS_NUM_THREADS` and the value passed to the `-n` flag of pytest with the `pytest-xdist` plugin;
- clean up unused files from the travis config (in a dedicated PR, once #20958 is merged).

cmarmo commented:
Thanks @ogrisel for your clarifications. I'm sorry to say that I will not have enough bandwidth to follow-up on this PR then.
Feel free to close it and choose the path you prefer for the next step.

ogrisel commented:
I light of the security [debacle at travis](https://twitter.com/peter_szilagyi/status/1437646118700175360) we have no choice but to move forward with completely migrating all the travis ARM builds to circle ci. And we will probably have to pay for the service to increase the concurrency a bit to make it fast enough when releasing.

alfaro96 commented:
> I light of the [recent security debacle at travis and their poor way of handling the matter](https://twitter.com/peter_szilagyi/status/1437646118700175360)

:upside_down_face:

ogrisel commented:
I think the first step for this plan would be to migrate the doc building jobs out of circle ci, for instance, to a new workflow on github actions.

This workflow could run:
- in "diff mode" in PRs (as done today)
- in full build mode using the `[doc build]` commit flag in PRs as done today
- in full build mode on main with upload to dev using a nightly schedule trigger (instead of at each merge commit on `main`)
- in full build mode on `main` with upload to `/dev` or on release branches with upload to `/X.Y` with a manual trigger in the github actions UI.

This way we should free the overworked circle ci executor to be able to focus on ARM builds. And if we decide to buy some circle ci build credits, they will be more wisely allocated to the scarce ARM building resources rather than being wasted on doc building which can be done more efficiently on github actions.

cmarmo commented:
Just linking here the CircleCI [documentation](https://circleci.com/docs/2.0/using-arm) about the improved Arm execution environment.... if this can be useful.

jjerphan commented:
Thank you for the pointer, @cmarmo. Do you want and/or have time to pursue this PR?

cmarmo commented:
> Thank you for the pointer, @cmarmo. Do you want and/or have time to pursue this PR?

@jjerphan , to be honest it is not clear to me if this feature is wanted or not... in general, I seem to have a different approach to CI pull requests with respect to what is expected by historical contributors.
I guess this PR is starting to be obsolete anyway.
I was just pointing out that, if performances were the issue, perhaps the new CircleCI Arm infrastructure could be helpful.

cmarmo commented:
I am closing this one as obsolete.
Also, it looks like Cirrus is the new black (#25048).
